### PR TITLE
Automaticaly track elements available from uncovered presence tracks.

### PR DESCRIPTION
**Pull request by tomprince**
- [x] Fix spirit validator: for some reason, it currently expands the scan zone a ridiculous amount, which overlaps Finder/Many Minds. I've not yet figured out why, though I suspect it may have something to do with the new button.
- [ ] Find the right place to but the new button, probably to the right of the scan zone to match the power/blight editors
- [x] Update all the spirits with appropriate data.
- [x] Consider refactoring `elemCombine`.
- [ ] Consider what to do about `Any` elements. I suspect it might make sense to pre-place them on the spirit board, underneath the presence.

### CODE REVIEW

**Comment by tomprince** [comment]
I've now updated all the spirits with the new data, and spot checked that a few of them are correct. I'm also happy with the refactoring to include this new logic.

I've moved the editor code into #75 as it needs a bit of UI polish, although it is usable now (I used to to add all the data for the spirits).

File: script.lua [hunk 1]
```diff
@@ -3226,57 +3225,96 @@ function setupPlayerArea(params)
...
+        return cost
+    end
+
+    local function calculateTrackElements(spiritBoard)
+        local elements = Elements:new()
+        if spiritBoard.script_state ~= "" then
+            local trackElements = spiritBoard.getVar("trackElements")
+            if trackElements ~= nil then
+                for _, trackElem in pairs(trackElements) do
+                    local hits = upCastPosSizRot(
```
Comment (iakona) [inline script.lua:3271]: this should use ray casting

File: script.lua [hunk 2]
```diff
@@ -3226,57 +3225,96 @@ function setupPlayerArea(params)
...
+                        spiritBoard.positionToWorld(trackElem.position), -- pos
+                        Vector(1,1,1),
+                        Vector(0,0,0),
+                        2
+                    )
+                    local hasPresence = false
+                    for _, hit in pairs(hits) do
+                        if hit.hasTag("Presence") then
+                            hasPresence = true
```
Comment (iakona) [inline script.lua:3280]: you should break out of for loop after you found presence

File: script.lua [hunk 3]
```diff
@@ -3226,57 +3225,96 @@ function setupPlayerArea(params)
...
-        for i = 1, string.len(elemStr) do
-            table.insert(outArr,(math.floor(string.sub(elemStr, i, i))))
-        end
-        return outArr
-    end
-
-    local function elemCombine(inTableOfElemStrCards)
+    local Elements = {}
+    Elements.__index = Elements
+    function Elements:new(init)
```
Comment (iakona) [inline script.lua:3230]: from what i can tell init is never passed in here, can that part just be removed?

Comment (tomprince) [reply script.lua:3230]: It is used in #75.

File: script.lua [hunk 4]
```diff
@@ -3230,57 +3230,97 @@ function setupPlayerArea(params)
...
+    local function calculateTrackElements(spiritBoard)
+        local elements = Elements:new()
+        if spiritBoard.script_state ~= "" then
+            local trackElements = spiritBoard.getVar("trackElements")
+            if trackElements ~= nil then
+                for _, trackElem in pairs(trackElements) do
+                    local hits = Physics.cast{
+                        origin = spiritBoard.positionToWorld(trackElem.position), -- pos
+                        direction = Vector(0, 1, 0),
+                        max_distance = 1,
```
Comment (iakona) [inline script.lua:3279]: max_distance of 1 seems a bit high, can this value be lower?

Comment (tomprince) [reply script.lua:3279]: I don't think it is worth optimizing this.

**Comment by iakona** [comment]
you're missing thunderspeaker (e9c468)

**Comment by tomprince** [comment]
Apparently I missed adding that file to VCS.

File: script.lua [hunk 5]
```diff
@@ -3229,57 +3229,97 @@ function setupPlayerArea(params)
...
-            local elemTable = elemStrToArr(inTableOfElemStrCards[i].getVar("elements"))
+        setmetatable(outTable, self)
+        outTable:add(init)
+        return outTable
+    end
+    function Elements:add(other)
+        if other == nil then
+            return
+        elseif type(other) == "table" then
             for j = 1, 8 do
```
Comment (iakona) [inline script.lua:3244]: the other conditional block uses i as a variable, for consistency this should also be i

Comment (iakona) [reply script.lua:3244]: eh on second thought i can just make a couple formatting changes on my end after the fact

**Review by iakona — APPROVED** [review]

**Pull request merged**